client/shared/src/main/scala/org/http4s/client/middleware/History.scala

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2014 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.client.middleware
+
+import cats.effect._
+import cats.implicits.catsSyntaxApplicativeId
+import org.http4s._
+import org.http4s.client.Client
+import org.http4s.headers.Date
+
+/** This Middleware provides history tracking of Uri's visited.
+  * History entries are ordered most recent to oldest.
+  *
+  * @param client: Client[F]
+  * @param history: Ref[F, Vector[HistoryEntry]
+  * @param maxSize: Int
+  */
+
+class HistoryEntry private (val httpDate: HttpDate, val method: Method, val uri: Uri) {
```

Comment:
A controversial opinion maybe, but I like to mark every class `final` whenever it makes sense

Reply:
The reason this isn't a case class is to reserve space for bincompat evolution.  Would we get more flexibility making it a sealed trait with a single private implementation?   I can't recall all the ramifications of adding a private impl later.

But if not sealed, yes, private.

Reply:
made HistoryEntry final, thanks